src/AutoMapper/Execution/TypeMapPlanBuilder.cs

```diff
@@ -1,309 +1,280 @@
-﻿namespace AutoMapper.Execution
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+
+namespace AutoMapper.Execution
 {
-    using System;
-    using System.Collections.Generic;
-    using System.Linq;
-    using System.Linq.Expressions;
-    using System.Reflection;
     using Configuration;
     using Mappers;
-    using static System.Linq.Expressions.Expression;
+    using static Expression;
     using static ExpressionExtensions;
 
-    public static class TypeMapPlanBuilder
-    {
-        public static LambdaExpression BuildMapperFunc(TypeMap typeMap, IConfigurationProvider configurationProvider, TypeMapRegistry typeMapRegistry)
-        {
-            if (typeMap.SourceType.IsGenericTypeDefinition() || typeMap.DestinationTypeToUse.IsGenericTypeDefinition())
+    public class TypeMapPlanBuilder
+    {
+        private static readonly Expression<Func<AutoMapperMappingException>> CtorExpression = () => new AutoMapperMappingException(null, null, default(TypePair), null, null);
+        private static readonly Expression<Action<ResolutionContext>> IncTypeDepthInfo = ctxt => ctxt.IncrementTypeDepth(default(TypePair));
+        private static readonly Expression<Action<ResolutionContext>> DecTypeDepthInfo = ctxt => ctxt.DecrementTypeDepth(default(TypePair));
+        private static readonly Expression<Func<ResolutionContext, int>> GetTypeDepthInfo = ctxt => ctxt.GetTypeDepth(default(TypePair));
+
+        readonly IConfigurationProvider _configurationProvider;
+        readonly TypeMap _typeMap;
+        readonly TypeMapRegistry _typeMapRegistry;
+        readonly ParameterExpression _source;
+        readonly ParameterExpression _initialDestination;
+        readonly ParameterExpression _context;
+        bool _constructorMapping;
```

Comment:
Small thing - either these fields should be readonly, or parameters, but not mutable fields. It's not consistent that some are readonly fields, some are mutable, and some are parameters.